### DISCUSSION TITLE
How to use local function in an index creation task in c#

### DISCUSSION THREAD

**Discussion by anemitoff** [ravendb/ravendb#19283] [Q&A] [answered]
I am trying to write a little function for "extractClaim" assignment in this index definition and I am getting an IndexCompilationException with "error CS0828: Cannot assign 'lambda expression' to anonymous type property"
```
AddMap<Record>(records =>
  from record in records
  let extractClaim = (Func<string,string?>)(name => record.claims.FirstOrDefault(x=>x.Key == name).Value)
  select new { 
     Name = extractClaim("name") ?? extractClaim("altname") 
  }
}
```

**Comment by Lwiel** [comment, marked as answer]
Is there any specific reason you want to have your function declared this way?

The usual way of using custom functions in the index definition is to define them in [additional sources](https://ravendb.net/docs/article-page/6.0/csharp/indexes/extending-indexes).

**Reply by anemitoff** [reply]
I forgot about "additional sources" and missed seeing anything about it in the documentation about writing static indexes that I had looked at while trying to find a solution.

I tried my approach because I am used to writing quick lambda expression helpers. Additional sources will work, but having the source be a string which is not subject to static runtime checking in my IDE is less than ideal.

**Reply by ayende** [reply]
Note you can register the file _twice_ once as a source file (checked) and another as embedded resource

**Reply by DejanMilicic** [reply]
Here is an example https://github.com/DejanMilicic/RavenWorkshop/blob/main/Northwind/Northwind/Features/AdditionalSources/AveragePrice/MyProducts_ByAveragePrice.cs#L37-L39
